### ISSUE TITLE
Faster hashing via non-crytographic functions

### ISSUE THREAD

**Issue by hmaarrfk** [fatiando/pooch#208]
It might be nice not to use cryptographic (slow) hashes to speed up comparison for large files.

This is especially true for large files (1GB) where it might take a long time to compute the SHA256.

Would you accept PRs that work toward taking a custom constructor for the hasher?

Proof of concept code below:
1. No speedup for mmap usage
2. Results
```python
In [30]: %time file_hash('mylarge_file, 'xxhash')
CPU times: user 81.4 ms, sys: 99.8 ms, total: 181 ms
Wall time: 180 ms
'2f3134bc224180353cbe73f2ba9e108b'

In [31]: %time file_hash('mylarge_file')
CPU times: user 1.48 s, sys: 72.1 ms, total: 1.55 s
Wall time: 1.55 s
'b87f472e26e694a5e991fc2c40b3b587c2db02500199490527d58874b6be1d7f'

```

```python
from mmap import mmap
import hashlib

def mmap_hash(fname, alg='sha256'):
    if alg == 'xxhash':
        import xxhash
        hasher = xxhash.xxh128()
    else:
        if alg not in hashlib.algorithms_available:
            raise ValueError("Algorithm '{}' not available in hashlib".format(alg))
        hasher = hashlib.new(alg)

    # Calculate the hash in chunks to avoid overloading the memory
    chunksize = 65536
    with open(fname, "r+b") as fin:
        # map the whole file
        mm = mmap(fin.fileno(), 0)
        file_size = len(mm)
        start_index = 0
        while start_index < file_size:
            end_index = start_index + chunksize
            hasher.update(mm[start_index:end_index])
            start_index = end_index
    return hasher.hexdigest()


def file_hash(fname, alg='sha256'):
    if alg == 'xxhash':
        import xxhash
        hasher = xxhash.xxh128()
    else:
        if alg not in hashlib.algorithms_available:
            raise ValueError("Algorithm '{}' not available in hashlib".format(alg))
        hasher = hashlib.new(alg)

    # Calculate the hash in chunks to avoid overloading the memory
    chunksize = 65536
    with open(fname, "r+b") as fin:
        buff = fin.read(chunksize)
        while buff:
            hasher.update(buff)
            buff = fin.read(chunksize)
    return hasher.hexdigest()
```

**Comment by hmaarrfk** [comment]
One challenge in implementing this is that `known_hash` doesn't take a custom `hash_algorithm` function.

Which means that it requires some updating to support specifying which hash is to be used with which file.

```
def hash_matches(fname, known_hash, strict=False, source=None):
```

**Comment by hmaarrfk** [comment]
^^^ I really should have read the example before posting this. You do support custom `hash_matches` has the required support, but `file_hash` needs to be able to take new constructors.

**Comment by leouieda** [comment]
@hmaarrfk that sounds like a great addition! The basis for this support was added by @Xarthisius in #133. You're right that SHA256 would be too slow for larger files. It wouldn't be a problem if we only did this on download but we run the hash check all the time.

One way to implement this would be:

* Choose identifiers for the xxhash methods. How about be `xxh32`, `xxh64` etc. 
* Modify [`file_hash`](https://github.com/fatiando/pooch/blob/16ee864d0980cfe39d45694dfac7f873db84c00b/pooch/utils.py#L65) to accept these identifiers as the `alg` argument and use `xxhash` instead of `hashlib` 
* Add optional dependency on `xxhash`. This would need to be added to the install page, [`requirements-optional.txt`](https://github.com/fatiando/pooch/blob/master/requirements-optional.txt) (so it gets run on CI), and flagged on the tests to only run when `xxhash` is available.
* Add instructions to https://www.fatiando.org/pooch/latest/beginner.html#hashes
